logical: add eval logical replication options

Jeremyyang920 opened the pull request:
This commit adds the ability for the plan hook to be able to evaluate the incoming options from the sql statement. As of right now, the function will only evaluate the options, but nothing will be done with them. If there is an error in the format of an option, the associated error will be returned.

See individual commits for details.

Epic: None
Release note: None

Commit:
logical: add eval replication options

This commit adds the ability for the plan hook to be able
to evaluate the incoming options from the sql statement. As
of right now, the function will only evaluate the options,
but nothing will be done with them. If there is an error in
the format of an option, the associated error will be returned.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/126520)
<!-- Reviewable:end -->

dt commented on pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:
```diff
@@ -86,8 +89,12 @@ func createLogicalReplicationStreamPlanHook(
-
 		if !stmt.Options.IsDefault() {
+			// TODO: update handling of the the returned options
+			_, err := evalLogicalReplicationOptions(ctx, stmt.Options, exprEval, &p.ExtendedEvalContext().Context, p.SemaCtx())
```
This is durinv exec; don’t we also need to do something during typecheck ?

Jeremyyang920 replied:
I think we would really only need 
```
	toTypeCheck := []exprutil.ToTypeCheck{
		exprutil.Strings{stmt.PGURL},
		exprutil.Strings{
			stmt.Options.Cursor,
			stmt.Options.DefaultFunction,
			stmt.Options.Mode,
		},
	}
```
WDYT? A similar eval check happens inside evalLogicalReplicationOptions when eval.String is called too.

Commit:
logical: add ability to use cursor

This commit adds the ability for the cursor option to be
plumbed into the job proto if it is supplied in the sql
syntax options.

Epic: None
Release note: None

Commit:
logical: parse udfs/default function options

This commit parses the default function clause
as well as the custom table to udf mapping.

We map the table name to the UDF descriptor ID and store
it inside the existing replication pairs map if the function
name can be resolved. We also now parse the default function
where LWW is the default if not set. If a UDF is provided, it
peforms the same lookup and stores the descriptor ID on the
new DefaultConflictResolution message on the jobs.pb proto.

Epic: None
Release note: None

Jeremyyang920 commented on pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:
```diff
@@ -213,11 +241,143 @@ func createLogicalReplicationStreamTypeCheck(
+	return r, nil
+}
+
+func lookupFunctionID(
+	ctx context.Context, p sql.PlanHookState, u tree.UnresolvedName,
+) (int32, error) {
+	rf, err := p.ResolveFunction(ctx, tree.MakeUnresolvedFunctionName(&u), &p.SessionData().SearchPath)
+	if err != nil {
+		return 0, err
+	}
+	if len(rf.Overloads) > 1 {
+		return 0, errors.Newf("function '%s' has more than 1 overload", u.String())
+	}
+	fnOID := rf.Overloads[0].Oid
+	descID := typedesc.UserDefinedTypeOIDToID(fnOID)
+	return int32(descID), nil
+}
```
@stevendanna So this lookup works based on our slack conversation. 
```
demo@127.0.0.1:26257/demoapp/movr> CREATE LOGICAL REPLICATION STREAM FROM TABLE users ON                                                                                                 
                                -> 'postgresql://demo:demo13904@127.0.0.1:26258/movr?options=-ccluster%3Ddemoapp&sslmode=require&sslrootcert=%2FUsers%2Fjeremyyang%2F.cockroach-demo%2Fca
                                -> .crt' INTO TABLE users WITH FUNCTION add FOR TABLE users;                                                                                             
[]jobspb.LogicalReplicationDetails_ReplicationPair{jobspb.LogicalReplicationDetails_ReplicationPair{SrcDescriptorID:106, DstDescriptorID:106, SrcFunctionID:112}}
```

Its now plumbed into the existing replication pairs details that already exists. Does that look useable for the UDF work you have?

stevendanna commented on pkg/ccl/crosscluster/logical/logical_replication_job_test.go:
```diff
@@ -202,6 +202,90 @@ func TestLogicalStreamIngestionJob(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestControlsTenantsExplicitly,
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+			},
+		},
+	}
+
+	server := testcluster.StartTestCluster(t, 1, clusterArgs)
+	defer server.Stopper().Stop(ctx)
+	s := server.Server(0).ApplicationLayer()
+
+	_, err := server.Conns[0].Exec("SET CLUSTER SETTING physical_replication.producer.timestamp_granularity = '0s'")
+	require.NoError(t, err)
+	_, err = server.Conns[0].Exec("CREATE DATABASE a")
+	require.NoError(t, err)
+	_, err = server.Conns[0].Exec("CREATE DATABASE B")
+	require.NoError(t, err)
+
+	dbA := sqlutils.MakeSQLRunner(s.SQLConn(t, serverutils.DBName("a")))
+	dbB := sqlutils.MakeSQLRunner(s.SQLConn(t, serverutils.DBName("b")))
+
+	for _, s := range testClusterSettings {
+		dbA.Exec(t, s)
+	}
+
+	createStmt := "CREATE TABLE tab (pk int primary key, payload string)"
+	dbA.Exec(t, createStmt)
+	dbB.Exec(t, createStmt)
+	dbA.Exec(t, "ALTER TABLE tab "+lwwColumnAdd)
+	dbB.Exec(t, "ALTER TABLE tab "+lwwColumnAdd)
```
We now have a helper function that will do this initial setup.

stevendanna commented on pkg/ccl/crosscluster/logical/logical_replication_job_test.go:
```diff
@@ -202,6 +202,90 @@ func TestLogicalStreamIngestionJob(t *testing.T) {
+	dbB.Exec(t, "INSERT INTO tab VALUES (3, 'celeriac')")
+	dbA.Exec(t, "UPSERT INTO tab VALUES (1, 'hello, again')")
+	dbB.Exec(t, "UPSERT INTO tab VALUES (1, 'goodbye, again')")
+	// We should expect starting at the provided now() to replicate all the data.
```
This test might be more interesting if you put `now` after some of the updates so that we can see it doesn't replicate things from before the option.  If that doesn't work yet, perhaps put it as a TODO.

Jeremyyang920 replied:
Adding as a TODO since it seems it doesn't work right now and am creating an issue that David said he can take a look at.

Jeremyyang920 replied:
Nevermind,  upon further checks, was just checking the wrong values. The test has been updated now.

Jeremyyang920 commented:
TFTR!

bors r+
